src/clients/HttpTunneler.cc

```diff
@@ -341,13 +351,49 @@ Http::Tunneler::bailWith(ErrorState *error)
 {
     Must(error);
     answer().squidError = error;
+
+    if (CachePeer *p = connection->getPeer())
+        peerConnectFailed(p);
+
     callBack();
+    disconnect();
+
+    if (noteFwdPconnUse)
+        fwdPconnPool->noteUses(fd_table[connection->fd].pconn.uses);
+    // TODO: Reuse to-peer connections after a CONNECT error response.
+    connection->close();
+    connection = nullptr;
+}
+
+void
+Http::Tunneler::sendSuccess()
+{
+    assert(answer().positive());
+    callBack();
+    disconnect();
+}
+
+void
+Http::Tunneler::disconnect()
+{
+    if (!Comm::IsConnOpen(connection))
```

Comment:
> We must reach that disconnected state even if the connection is already closing.

The most possible is that the disconnect name is not good, we should rename it to be "stopMonitoringServer" or something like that.  Because we are not actually disconnecting from connection here, the connection object may continue used by Tunneler after this call is called.

Please also check my comments bellow for PeerConnector::disconnect method. More or less are valid for this method to.

Reply:
> we are not actually disconnecting from connection here

We are disconnecting from Comm (i.e. making sure that Comm does not call our job). I am comfortable with the current method name, but am happy to consider better alternatives. For example, if you would prefer disconnectFromComm(), please use that instead.

Reply:
I think this is resolved with commit a378a06